components/script/window_named_properties.rs

```diff
@@ -0,0 +1,263 @@
+            isExtensible: Some(is_extensible),
+            has: None,
+            get: None,
+            set: None,
+            call: None,
+            construct: None,
+            hasOwn: None,
+            getOwnEnumerablePropertyKeys: None,
+            nativeCall: None,
+            objectClassIs: None,
+            className: Some(class_name),
+            fun_toString: None,
+            boxedValue_unbox: None,
+            defaultValue: None,
+            trace: None,
+            finalize: None,
+            objectMoved: None,
+            isCallable: None,
+            isConstructor: None,
+        };
+
+        #[allow(unsafe_code)]
+        unsafe {
+            SyncWrapper(CreateProxyHandler(&traps, ptr::null()))
+        }
+    };
+}
+
+#[allow(unsafe_code)]
+unsafe extern "C" fn get_own_property_descriptor(
+    cx: *mut JSContext,
+    proxy: HandleObject,
+    id: HandleId,
+    desc: MutableHandle<PropertyDescriptor>,
+    is_none: *mut bool,
+) -> bool {
+    let cx = SafeJSContext::from_ptr(cx);
+
+    if id.is_symbol() {
+        if id.get().asBits_ == SymbolId(GetWellKnownSymbol(*cx, SymbolCode::toStringTag)).asBits_ {
+            rooted!(in(*cx) let mut rval = UndefinedValue());
+            "WindowProperties".to_jsval(*cx, rval.handle_mut());
+            set_property_descriptor(
+                RustMutableHandle::from_raw(desc),
+                rval.handle(),
+                0,
+                &mut *is_none,
+            );
+        }
+        return true;
+    }
+
+    let mut found = false;
+    if !has_property_on_prototype(
+        *cx,
+        RustHandle::from_raw(proxy),
+        RustHandle::from_raw(id),
+        &mut found,
+    ) {
+        return false;
+    }
+    if found {
+        return true;
+    }
+
+    let s = if id.is_string() {
+        jsstr_to_string(*cx, id.to_string())
+    } else if id.is_int() {
+        // If the property key is an integer index, convert it to a String too.
+        // TODO(delan) will this interfere with indexed access on the Window object
+        // (window[index]), which should only return document-tree child navigables?
+        // https://html.spec.whatwg.org/#accessing-other-browsing-contexts
+        id.to_int().to_string()
+    } else if id.is_symbol() {
+        unreachable!()
+    } else {
+        unreachable!()
+    };
+    if s.is_empty() {
+        return true;
+    }
+
+    let window = Root::downcast::<Window>(GlobalScope::from_object(proxy.get()))
+        .expect("global is not a window");
+    if let Some(obj) = window.NamedGetter(cx, s.into()) {
+        rooted!(in(*cx) let mut rval = UndefinedValue());
+        obj.to_jsval(*cx, rval.handle_mut());
+        set_property_descriptor(
+            RustMutableHandle::from_raw(desc),
+            rval.handle(),
+            0,
+            &mut *is_none,
+        );
+    }
+    return true;
+}
+
+#[allow(unsafe_code)]
+unsafe extern "C" fn own_property_keys(
```

Comment:
https://searchfox.org/mozilla-central/rev/af78418c4b5f2c8721d1a06486cf4cf0b33e1e8d/dom/base/WindowNamedPropertiesHandler.cpp#175-232 makes me believe that we should be returning more items here. Want to file an issue to keep track of this?

Reply:
looking into this…

Reply:
i’ve spent hours trying to figure out who’s correct here, but i’m still not sure, so i’ve opened whatwg/html#9068

Reply:
added comment in 557ce73f5f8acf7f97c49ff0e50c8cdc8b0f178d